subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c

```diff
@@ -0,0 +1,1222 @@
+				 rx_conf->block_count * rx_conf->block_size;
+	size_t name_size;
+	size_t i;
+
+	/* Requested name is in shared memory, so we have to assume that it
+	 * can be corrupted. Extra care must be taken to avoid out of
+	 * bounds reads.
+	 */
+	name_size = strnlen(name, buffer_end - name - 1) + 1;
+
+	for (i = 0; i < dev_data->ept_count; i++) {
+		if (strncmp(dev_data->ept[i].cfg->name, name, name_size) == 0) {
+			return &dev_data->ept[i];
+		}
+	}
+	return NULL;
+}
+
+/**
+ * Find registered endpoint that matches given "bound endpoint" message. When found,
+ * remote endpoint address is saved for future use and "release bound endpoint" message
+ * is send. This function must be called when mutex is locked.
+ *
+ * @param[in] rx_block_index	Block containing the "bound endpoint" message.
+ *
+ * @retval 0	match not found
+ * @retval 1	match found and processing was successful
+ * @retval < 0	error occurred during processing
+ */
+static int match_bound_msg(struct backend_data *dev_data, size_t rx_block_index)
+{
+	const struct icmsg_with_buf_config *conf = dev_data->conf;
+	uint8_t *buffer;
+	struct ept_bound_msg *msg;
+	struct ept_data *ept;
+	struct block_header *block;
+	int r = 0;
+
+	block = block_from_index(&conf->rx, rx_block_index);
+	buffer = block->data;
+	msg = (struct ept_bound_msg *)buffer;
+
+	ept = find_ept_by_name(dev_data, msg->name);
+
+	if (ept == NULL) {
+		return 0;
+	}
+
+	ept->remote_addr = msg->ept_addr;
+
+	k_mutex_unlock(&dev_data->mutex);
+	r = send_release(dev_data, buffer, MSG_RELEASE_BOUND);
+	k_mutex_lock(&dev_data->mutex, K_FOREVER);
+
+	if (r < 0) {
+		return r;
+	}
+
+	return 1;
+}
+
+/**
+ * Send bound message on specified endpoint.
+ *
+ * @param[in] ept	Endpoint to use
+ *
+ * @return		O or negative error code
+ */
+static int send_bound_message(struct backend_data *dev_data, struct ept_data *ept)
+{
+	size_t msg_len;
+	uint32_t alloc_size;
+	uint8_t *buffer;
+	struct ept_bound_msg *msg;
+	int r;
+
+	msg_len = offsetof(struct ept_bound_msg, name) + strlen(ept->cfg->name) + 1;
+	alloc_size = msg_len;
+	r = alloc_tx_buffer(dev_data, &alloc_size, &buffer, K_FOREVER);
+	if (r >= 0) {
+		msg = (struct ept_bound_msg *)buffer;
+		msg->ept_addr = ept->local_addr;
+		strcpy(msg->name, ept->cfg->name);
+		r = send_block(dev_data, r, msg_len, MSG_BOUND);
+	}
+
+	return r;
+}
+
+/**
+ * Put endpoint bound processing into system workqueue.
+ */
+static void schedule_ept_bound_process(struct backend_data *dev_data)
+{
+	k_work_submit(&dev_data->ep_bound_work);
+}
+
+/**
+ * Work handler that is responsible for bounding the endpoints. It sends endpoint
+ * bound message to the remote and calls local endpoint bound callback.
+ */
+static void ept_bound_process(struct k_work *item)
+{
+	struct backend_data *dev_data = CONTAINER_OF(item, struct backend_data,
+						     ep_bound_work);
+	struct ept_data *ept = NULL;
+	size_t i;
+	int r = 0;
+
+	k_mutex_lock(&dev_data->mutex, K_FOREVER);
```

Comment:
Maybe configurable timeout could be an option?

Reply:
I disagree. It is only to avoid access to the same memory by more threads. It is not long running procedure, so in such case, I don't see point of using something different then K_FOREVER.